Add h3Line, v3.3.0

isaacbrodsky opened the pull request:
Fixes https://github.com/uber/h3/issues/185

The tests for h3Line are based on the ones in @uber/h3-js. I used LocalIjUndefinedException, but DistanceUndefinedException or a new LineUndefinedException might make sense.

v3.4.0 will be a separate PR.

nrabinowitz commented on src/main/java/com/uber/h3core/H3Core.java:
```diff
@@ -565,6 +565,26 @@ public String experimentalLocalIjToH3(String originAddress, CoordIJ ij) throws L
+        int size = h3Api.h3LineSize(start, end);
+
+        if (size < 0) {
+            throw new LocalIjUndefinedException("Could not compute line between cells");
```
I'd prefer `LineUndefinedException` here and below. This function doesn't have an internal knowledge of local IJ coords, and it's entirely possible that we'd change the underlying implementation of `h3Line` to not use local IJ coords - if we have to change the exception here when we make that change, we turn an invisible, internal update into a major version change.

isaacbrodsky replied:
I think that's probably the right approach. I'll update the PR.

nrabinowitz commented on src/test/java/com/uber/h3core/TestH3Core.java:
```diff
@@ -961,4 +961,30 @@ public void testExperimentalLocalIjToH3Pentagon() throws LocalIjUndefinedExcepti
+
+    @Test
+    public void testH3Line() throws LocalIjUndefinedException, DistanceUndefinedException {
+        for (int res = 0; res < 5; res++) {
```
Nit: 5 is pretty arbitrary; I just bumped this to 12 (15 is possible but might have an impact on test run time).

isaacbrodsky commented:
I updated the PR to add coverage for the case that `h3Line` fails - but to reach that I would need indexes that succeed in `h3Distance` when called from `h3LineSize` but when called from `h3Line`. There shouldn't be any indexes that do that, so I modified the Java library to permit injecting the failure case.

The other options I can see are to add a package private that makes the call to `h3Line` but doesn't call `h3LineSize` (doesn't seem to have an advantage here), use reflection to instantiate `H3Core` rather than promoting `private H3Core(NativeMethods)`, or leave the line uncovered.

nrabinowitz commented:
> I updated the PR to add coverage for the case that `h3Line` fails - but to reach that I would need indexes that succeed in `h3Distance` when called from `h3LineSize` but when called from `h3Line`. 

Per offline conversation, I'm happy leaving this line uncovered (ideally you can mark it for coverage exclusion). It doesn't seem worthwhile going through contortions to cover a line that's expected to be unreachable.
